README.md
# django-health-check-plus


CHANGELOG
Changes
=======
v2.0.0 - 13/05/2021
* Update required `django-health-check` to v3.16.4 so it could support Django 3.

v1.0.1 - 02/08/2018
* Fix optional check view not detecting custom health checks.

v1.0.0 - 11/08/2017
* Upgrade to Django 1.11, python3 and `django-health-check` 2.3.0
* [Breaking change] Removed old functionality and remove old Django version compatibility.


v0.2.0 - 08/06/2017
* JSonResponse Django native method will be used for 1.7 version
* JSonResponse method encoder changed to DjangoJSONEncoder
* Deprecated Mime-Type HTTP header changed to Content-Type header


v0.1.1 - 12/06/2015
 * Minor changes in packaging and readme.

v0.1.0 - 11/06/2015
 * Initial version

health_check_plus/tests.py
from django.conf import settings
from django.test import TestCase, override_settings
from django.urls import reverse

from health_check.backends import BaseHealthCheckBackend
from health_check.exceptions import ServiceUnavailable
from health_check.plugins import plugin_dir


class StatusOptionalCheckViewTest(TestCase):
    class ServiceDownCustomHealthCheck(BaseHealthCheckBackend):
        def check_status(self):
            raise ServiceUnavailable('Testing service not working')

    class ServiceUpCustomHealthCheck(BaseHealthCheckBackend):
        def check_status(self):
            pass

    class ServiceUpOverriddenIdentifierCustomHealthCheck(BaseHealthCheckBackend):
        def check_status(self):
            pass

        @staticmethod
        def identifier():
            return 'test_check'

    url = reverse('health_optional_check')

    def test_query_url_with_not_params_returns_html(self):
        result = self.client.get('{}?checks'.format(self.url))

        self.assertEqual(result.status_code, 200)
        with self.assertRaises(ValueError,
                               msg='Content-Type header is "text/html; charset=utf-8", not "application/json"'):
            result.json()

    def test_query_url_with_format_json_returns_json(self):
        result = self.client.get('{}?format=json&checks'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {})

    @override_settings(INSTALLED_APPS=settings.INSTALLED_APPS + ['health_check.db'])
    def test_checks_db_returns_db_status(self):
        result = self.client.get('{}?format=json&checks=db'.format(self.url))

        self.assertEqual(result.status_code, 500)
        self.assertEqual(result.json()['DatabaseBackend'], 'unavailable: Database error')

    @override_settings(INSTALLED_APPS=settings.INSTALLED_APPS + ['health_check.db'])
    def test_even_if_db_crash_checks_returns_ok(self):
        result = self.client.get('{}?format=json&checks=cache'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {})

    @override_settings(INSTALLED_APPS=settings.INSTALLED_APPS + ['health_check.db'])
    def test_if_no_checks_in_params_it_checks_all(self):
        result = self.client.get('{}?format=json'.format(self.url))

        self.assertEqual(result.status_code, 500)
        self.assertEqual(result.json()['DatabaseBackend'], 'unavailable: Database error')

    def test_custom_health_check_not_failing(self):
        plugin_dir.register(self.ServiceUpCustomHealthCheck)

        result = self.client.get('{}?format=json&checks=ServiceUpCustomHealthCheck'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {'ServiceUpCustomHealthCheck': 'working'})

    def test_custom_health_check_failing(self):
        plugin_dir.register(self.ServiceDownCustomHealthCheck)

        result = self.client.get('{}?format=json&checks=ServiceDownCustomHealthCheck'.format(self.url))

        self.assertEqual(result.status_code, 500)
        self.assertEqual(result.json(), {'ServiceDownCustomHealthCheck': 'unavailable: Testing service not working'})

    def test_custom_health_overridden_identifier_check_not_failing(self):
        plugin_dir.register(self.ServiceUpOverriddenIdentifierCustomHealthCheck)

        result = self.client.get('{}?format=json&checks=test_check'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {'test_check': 'working'})

    def test_two_custom_health_check_one_down_other_up(self):
        plugin_dir.register(self.ServiceUpCustomHealthCheck)
        plugin_dir.register(self.ServiceDownCustomHealthCheck)

        result = self.client.get(
            '{}?format=json&checks=ServiceUpCustomHealthCheck,ServiceDownCustomHealthCheck'.format(self.url))

        self.assertEqual(result.status_code, 500)
        self.assertEqual(
            result.json(),
            {
                'ServiceDownCustomHealthCheck': 'unavailable: Testing service not working',
                'ServiceUpCustomHealthCheck': 'working'
            }
        )


class StatusCheckPingViewTest(TestCase):
    url = reverse('health_check_ping')

    def test_query_url_with_returns_always_json(self):
        result = self.client.get('{}'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {})

    @override_settings(INSTALLED_APPS=settings.INSTALLED_APPS + ['health_check.db'])
    def test_even_if_db_check_enabled_does_not_matter_it_returns_200_anyway(self):
        result = self.client.get('{}'.format(self.url))

        self.assertEqual(result.status_code, 200)
        self.assertEqual(result.json(), {})


runtest.py
import django
import sys

from django.conf import settings
from django.test.runner import DiscoverRunner

settings.configure(
    DEBUG=True,
    SECRET_KEY='secret_key',
    ROOT_URLCONF='health_check_plus.urls',
    INSTALLED_APPS=[
       'django.contrib.auth',
       'django.contrib.contenttypes',
       'django.contrib.messages',
       'django.contrib.staticfiles',
       'health_check',
       'health_check_plus'],
    MIDDLEWARE=[],
    DATABASES={
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': 'sqlite3'
        }
    },
    ALLOWED_HOSTS=['*'],
    TEST_RUNNER='runtest.NoDbTestRunner',
    TEMPLATES=[
        {
            'BACKEND': 'django.template.backends.django.DjangoTemplates',
            'DIRS': [],
            'APP_DIRS': True,
            'OPTIONS': {
                'context_processors': [
                    'django.template.context_processors.debug',
                    'django.template.context_processors.request',
                    'django.contrib.auth.context_processors.auth',
                    'django.contrib.messages.context_processors.messages',
                ],
            },
        }
    ]
)

django.setup()
test_runner = DiscoverRunner(verbosity=1)

failures = test_runner.run_tests(['health_check_plus'])
if failures:
    sys.exit(failures)


health_check_plus/utils.py
import json

import django
from django.http import HttpResponse

DJANGO_17_MINOR_VERSION = 7

if django.VERSION[1] == DJANGO_17_MINOR_VERSION:
    from django.http import JsonResponse
else:
    from django.core.serializers.json import DjangoJSONEncoder


    class JsonResponse(HttpResponse):
        def __init__(self, data, encoder=DjangoJSONEncoder, safe=True, **kwargs):
            if safe and not isinstance(data, dict):
                raise TypeError('In order to allow non-dict objects to be '
                                'serialized set the safe parameter to False')
            if 'mimetype' in kwargs:
                raise TypeError('To keep code clean to Django 1.7 JsonResponse do '
                                'not allow mimetype argument since 1.7 version')

            kwargs.setdefault('content_type', 'application/json')
            data = json.dumps(data, cls=encoder)
            super(JsonResponse, self).__init__(content=data, **kwargs)


health_check_plus/views.py
import copy

from django.http import JsonResponse
from health_check.views import MainView
from health_check.plugins import plugin_dir


class StatusOptionalCheckView(MainView):
    humanize_plugin_name = {
        'DatabaseBackend': 'db',
        'CacheBackend': 'cache',
        'CeleryHealthCheck': 'celery',
        'DefaultFileStorageHealthCheck': 'storage'
    }

    def get(self, request, *args, **kwargs):
        plugins = []
        errors = []
        for plugin_class, options in plugin_dir._registry:
            plugin = plugin_class(**copy.deepcopy(options))
            if self._plugin_in_queryparams(request, plugin):
                plugin.run_check()
                plugins.append(plugin)
                errors += plugin.errors
        plugins.sort(key=lambda x: x.identifier())
        status_code = 500 if errors else 200

        if 'application/json' in request.META.get('HTTP_ACCEPT', '') or request.GET.get('format') == 'json':
            return self.render_to_response_json(plugins, status_code)

        return self.render_to_response({'plugins': plugins}, status=status_code)

    def _plugin_in_queryparams(self, request, plugin):
        if 'checks' in request.GET:
            plugins_to_check = request.GET.get('checks', '').split(',')
            return self.humanize_plugin_name.get(plugin.identifier(), plugin.identifier()) in plugins_to_check
        return True


class StatusCheckPingView(MainView):

    def get(self, request, *args, **kwargs):
        return JsonResponse({}, status=200)
